In [5]:
import pandas as pd
import altair as alt
import numpy as np

In [6]:
# run the following for the first time
# try:
#     %pip install --user "oracledb" --no-warn-script-location
# except Exception as e:
#     print("\x1b[31m\u2717 Unexpected error! Please contact course staff\n" +
#           "Please include the entire text above and below in your message.")
#     raise

In [7]:
import oracledb

In [25]:
dsn = oracledb.makedsn("localhost", 1522, service_name = "stu")
connection = oracledb.connect(user="ora_garora08", password = "a98097926", dsn=dsn)

In [26]:
cur = connection.cursor()
cur.execute("SELECT table_name FROM user_tables")
for row in cur:
    print(row)

('AUTHORS',)
('CLEAN_DATA_CIUS',)
('CLEAN_WHO_HAS_COVID',)
('EDITORS',)
('PUBLISHERS',)
('SALES',)
('SALESDETAILS',)
('TITLEAUTHORS',)
('TITLEDITORS',)
('TITLES',)


In [27]:
cur.execute("SELECT age_group, year, access_to_internet, instant_messaging, social_media, voice_calls, New_cases FROM CLEAN_DATA_CIUS ORDER BY age_group, year")
for row in cur:
    print(row)

('15 to 24 years', 2018, 99, 90, 91, 68, 0)
('15 to 24 years', 2020, 98, 93, 91, 80, 759)
('15 to 24 years', 2022, 99, 92, 91, 82, 27326)
('25 to 44 years', 2018, 98, 89, 86, 61, 0)
('25 to 44 years', 2020, 98, 91, 87, 80, 759)
('25 to 44 years', 2022, 99, 91, 90, 80, 27326)
('45 to 64 years', 2018, 94, 69, 63, 40, 0)
('45 to 64 years', 2020, 94, 75, 66, 59, 759)
('45 to 64 years', 2022, 96, 78, 72, 61, 27326)
('65 years and over', 2018, 71, 34, 32, 21, 0)
('65 years and over', 2020, 76, 44, 40, 37, 759)
('65 years and over', 2022, 83, 51, 44, 38, 27326)
('Total, 15 years and over', 2018, 91, 71, 69, 47, 0)
('Total, 15 years and over', 2020, 92, 76, 71, 64, 759)
('Total, 15 years and over', 2022, 95, 78, 74, 65, 27326)


In [28]:
cur.execute("SELECT * FROM CLEAN_DATA_CIUS")
for row in cur:
    print(row)

('15 to 24 years', 2018, 99, 92, 90, 91, 68, 11, 71, 88, None, None, None, 94, 36, 79, 92, 15, None, None, None, 34, 34, 58, 2, None, 36, 69, None, 22, 21, 38, 32, None, 62, 36, 87, 68, 74, 25, 13, 56, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 0)
('25 to 44 years', 2018, 98, 94, 89, 86, 61, 8, 84, 91, None, None, None, 84, 37, 74, 84, 14, None, None, None, 32, 31, 38, 3, None, 30, 90, None, 20, 21, 37, 49, None, 83, 34, 92, 70, 84, 29, 21, 76, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 0)
('45 to 64 years', 2018, 94, 88, 69, 63, 40, 3, 73, 83, None, None, None, 59, 20, 49, 61, 8, None, None, None, 23, 22, 20, 2, None, 17, 76, None, 8, 11, 21, 32, None, 71, 24, 78, 46, 67, 14, 12, 58, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 0)
('65 years and over', 2018, 71, 63, 34, 32, 21,

In [29]:
question2 = cur.execute("SELECT year, round(avg(access_to_internet), 3) as avg_internet_access, round(avg(social_media), 3) as avg_social_media_usage, round(avg(instant_messaging), 3) as avg_instant_messaging, round(avg(voice_calls), 3) as avg_voice_calls, round(avg(New_cases), 3) as avg_covid_cases FROM clean_data_CIUS GROUP BY year ORDER BY year")
for row in question2:
    print(row)

(2018, 90.6, 68.2, 70.6, 47.4, 0)
(2020, 91.6, 71, 75.8, 64, 759)
(2022, 94.4, 74.2, 78, 65.2, 27326)


In [30]:
question3 = cur.execute("SELECT a.age_group, a.access_to_internet as access_internet_2020, b.access_to_internet as access_internet_2022, round(((b.access_to_internet - a.access_to_internet) / a.access_to_internet) * 100, 3) as internet_usage_change_from_2020_to_2022  FROM clean_data_CIUS a JOIN clean_data_CIUS b ON a.age_group = b.age_group WHERE a.year = 2020 AND b.year = 2022")
for row in question3:
    print(row)

('15 to 24 years', 98, 99, 1.02)
('25 to 44 years', 98, 99, 1.02)
('45 to 64 years', 94, 96, 2.128)
('65 years and over', 76, 83, 9.211)
('Total, 15 years and over', 92, 95, 3.261)


In [31]:
query4 = cur.execute("SELECT a.age_group, a.voice_calls as voice_calls_2020, b.voice_calls as voice_calls_2022, round(((b.voice_calls - a.voice_calls) / a.voice_calls) * 100, 3) as voice_calls_change_from_2020_to_2022 FROM clean_data_CIUS a JOIN clean_data_CIUS b ON a.age_group = b.age_group WHERE a.year = 2020 AND b.year = 2022")
for row in query4:
    print(row)

('15 to 24 years', 80, 82, 2.5)
('25 to 44 years', 80, 80, 0)
('45 to 64 years', 59, 61, 3.39)
('65 years and over', 37, 38, 2.703)
('Total, 15 years and over', 64, 65, 1.563)


In [34]:
query5 = cur.execute("SELECT a.age_group, a.social_media as social_media_2020, b.social_media as social_media_2022, round(((b.social_media - a.social_media) / a.social_media) * 100, 3) as social_media_change_from_2020_to_2022 FROM clean_data_CIUS a JOIN clean_data_CIUS b ON a.age_group = b.age_group WHERE a.year = 2020 AND b.year = 2022")
for row in query5:
    print(row)

('15 to 24 years', 91, 91, 0)
('25 to 44 years', 87, 90, 3.448)
('45 to 64 years', 66, 72, 9.091)
('65 years and over', 40, 44, 10)
('Total, 15 years and over', 71, 74, 4.225)


In [37]:

#  DataFrames from the provided data
# Internet usage by year
internet_by_year_data = {
    'year': [2018, 2020, 2022],
    'access_to_internet': [90.6, 91.6, 94.4],
    'social_media': [68.2, 71, 74.2],
    'instant_messaging': [70.6, 75.8, 78],
    'voice_calls': [47.4, 64, 65.2],
    'covid_cases': [0, 759, 27326]
}
internet_by_year_df = pd.DataFrame(internet_by_year_data)

# Age group data
age_group_data = [
    {'age_group': '15 to 24 years', 'year': 2018, 'access_to_internet': 99, 'instant_messaging': 90, 'social_media': 91, 'voice_calls': 68, 'new_cases': 0},
    {'age_group': '15 to 24 years', 'year': 2020, 'access_to_internet': 98, 'instant_messaging': 93, 'social_media': 91, 'voice_calls': 80, 'new_cases': 759},
    {'age_group': '15 to 24 years', 'year': 2022, 'access_to_internet': 99, 'instant_messaging': 92, 'social_media': 91, 'voice_calls': 82, 'new_cases': 27326},
    {'age_group': '25 to 44 years', 'year': 2018, 'access_to_internet': 98, 'instant_messaging': 89, 'social_media': 86, 'voice_calls': 61, 'new_cases': 0},
    {'age_group': '25 to 44 years', 'year': 2020, 'access_to_internet': 98, 'instant_messaging': 91, 'social_media': 87, 'voice_calls': 80, 'new_cases': 759},
    {'age_group': '25 to 44 years', 'year': 2022, 'access_to_internet': 99, 'instant_messaging': 91, 'social_media': 90, 'voice_calls': 80, 'new_cases': 27326},
    {'age_group': '45 to 64 years', 'year': 2018, 'access_to_internet': 94, 'instant_messaging': 69, 'social_media': 63, 'voice_calls': 40, 'new_cases': 0},
    {'age_group': '45 to 64 years', 'year': 2020, 'access_to_internet': 94, 'instant_messaging': 75, 'social_media': 66, 'voice_calls': 59, 'new_cases': 759},
    {'age_group': '45 to 64 years', 'year': 2022, 'access_to_internet': 96, 'instant_messaging': 78, 'social_media': 72, 'voice_calls': 61, 'new_cases': 27326},
    {'age_group': '65 years and over', 'year': 2018, 'access_to_internet': 71, 'instant_messaging': 34, 'social_media': 32, 'voice_calls': 21, 'new_cases': 0},
    {'age_group': '65 years and over', 'year': 2020, 'access_to_internet': 76, 'instant_messaging': 44, 'social_media': 40, 'voice_calls': 37, 'new_cases': 759},
    {'age_group': '65 years and over', 'year': 2022, 'access_to_internet': 83, 'instant_messaging': 51, 'social_media': 44, 'voice_calls': 38, 'new_cases': 27326},
    {'age_group': 'Total, 15 years and over', 'year': 2018, 'access_to_internet': 91, 'instant_messaging': 71, 'social_media': 69, 'voice_calls': 47, 'new_cases': 0},
    {'age_group': 'Total, 15 years and over', 'year': 2020, 'access_to_internet': 92, 'instant_messaging': 76, 'social_media': 71, 'voice_calls': 64, 'new_cases': 759},
    {'age_group': 'Total, 15 years and over', 'year': 2022, 'access_to_internet': 95, 'instant_messaging': 78, 'social_media': 74, 'voice_calls': 65, 'new_cases': 27326}
]
age_group_df = pd.DataFrame(age_group_data)

# Changes from 2020 to 2022 data
change_data = [
    {'age_group': '15 to 24 years', 'internet_change': 1.02, 'social_media_change': 0, 'voice_calls_change': 2.5},
    {'age_group': '25 to 44 years', 'internet_change': 1.02, 'social_media_change': 3.448, 'voice_calls_change': 0},
    {'age_group': '45 to 64 years', 'internet_change': 2.128, 'social_media_change': 9.091, 'voice_calls_change': 3.39},
    {'age_group': '65 years and over', 'internet_change': 9.211, 'social_media_change': 10, 'voice_calls_change': 2.703},
    {'age_group': 'Total, 15 years and over', 'internet_change': 3.261, 'social_media_change': 4.225, 'voice_calls_change': 1.563}
]
change_df = pd.DataFrame(change_data)


internet_usage_melted = internet_by_year_df.melt(
    id_vars=['year'],
    value_vars=['access_to_internet', 'social_media', 'instant_messaging', 'voice_calls'],
    var_name='activity_type',
    value_name='percentage'
)

# 1. Overall internet usage trends vs COVID cases

base = alt.Chart(internet_by_year_df).encode(
    x=alt.X('year:O', title='Year')
)

# Line chart for internet metrics
lines = base.mark_line(point=True).encode(
    y=alt.Y('value:Q', title='Usage Percentage (%)'),
    color='key:N',
    strokeWidth=alt.value(3)
).transform_fold(
    ['access_to_internet', 'social_media', 'instant_messaging', 'voice_calls'],
    as_=['key', 'value']
).properties(
    title='Internet Usage vs COVID-19 Cases (2018-2022)'
)


internet_by_year_df['scaled_covid'] = internet_by_year_df['covid_cases'] / 400


bars = alt.Chart(internet_by_year_df).mark_bar(opacity=0.3, color='red').encode(
    x=alt.X('year:O'),
    y=alt.Y('scaled_covid:Q', title='COVID-19 Cases (hundreds)', axis=alt.Axis(titleColor='red'))
)

# Combine the charts
chart1 = alt.layer(lines, bars).resolve_scale(
    y='independent'
).properties(
    width=600,
    height=400
)

# 2. Internet access by age group over time
chart2 = alt.Chart(age_group_df[age_group_df['age_group'] != 'Total, 15 years and over']).mark_line(point=True).encode(
    x=alt.X('year:O', title='Year'),
    y=alt.Y('access_to_internet:Q', title='Internet Access (%)'),
    color=alt.Color('age_group:N', title='Age Group'),
    strokeWidth=alt.value(3)
).properties(
    title='Internet Access by Age Group (2018-2022)',
    width=600,
    height=400
)

# 3. Percent change in internet metrics by age group (2020-2022)
# Melt the change_df for easier plotting
change_melted = change_df[change_df['age_group'] != 'Total, 15 years and over'].melt(
    id_vars=['age_group'],
    value_vars=['internet_change', 'social_media_change', 'voice_calls_change'],
    var_name='metric',
    value_name='percent_change'
)

# Clean up the metric names for display
change_melted['metric'] = change_melted['metric'].replace({
    'internet_change': 'Internet Access',
    'social_media_change': 'Social Media',
    'voice_calls_change': 'Voice Calls'
})

chart3 = alt.Chart(change_melted).mark_bar().encode(
    x=alt.X('age_group:N', title='Age Group', sort=['15 to 24 years', '25 to 44 years', '45 to 64 years', '65 years and over']),
    y=alt.Y('percent_change:Q', title='Percent Change (%)'),
    color=alt.Color('metric:N', title='Metric'),
    column=alt.Column('metric:N', title='Changes from 2020 to 2022')
).properties(
    title='Percent Change in Internet Usage Metrics by Age Group (2020-2022)',
    width=150,
    height=300
)

# 4. Voice calls usage by age group (showing the dramatic pandemic increase)
chart4 = alt.Chart(age_group_df[age_group_df['age_group'] != 'Total, 15 years and over']).mark_line(point=True).encode(
    x=alt.X('year:O', title='Year'),
    y=alt.Y('voice_calls:Q', title='Voice Calls Usage (%)'),
    color=alt.Color('age_group:N', title='Age Group'),
    strokeWidth=alt.value(3)
).properties(
    title='Voice Calls Usage by Age Group (2018-2022)',
    width=600,
    height=400
)

# 5. Heatmap showing the percentage changes across age groups and metrics
# Prepare data for the heatmap
heat_df = change_df[change_df['age_group'] != 'Total, 15 years and over'].copy()
heat_melted = heat_df.melt(
    id_vars=['age_group'],
    value_vars=['internet_change', 'social_media_change', 'voice_calls_change'],
    var_name='metric',
    value_name='percent_change'
)

# Clean up the metric names for display
heat_melted['metric'] = heat_melted['metric'].replace({
    'internet_change': 'Internet Access',
    'social_media_change': 'Social Media',
    'voice_calls_change': 'Voice Calls'
})

chart5 = alt.Chart(heat_melted).mark_rect().encode(
    x=alt.X('age_group:N', title='Age Group', sort=['15 to 24 years', '25 to 44 years', '45 to 64 years', '65 years and over']),
    y=alt.Y('metric:N', title='Metric'),
    color=alt.Color('percent_change:Q', scale=alt.Scale(scheme='blues'), title='Percent Change (%)'),
    tooltip=['age_group', 'metric', 'percent_change']
).properties(
    title='Heatmap of Percent Changes in Internet Usage (2020-2022)',
    width=500,
    height=200
)

# Display charts
chart1 & chart2 & chart3 & chart4 & chart5

alt.VConcatChart(...)

## Analysis Conclusions

### COVID-19 Impact on Internet Use:
- There is a clear correlation between rising COVID-19 cases and increased internet usage across all metrics.
- **Voice calls** saw the most dramatic increase, rising from **47.4% to 64%** during the initial pandemic phase (2018-2020).

### Persistence of Pandemic Habits:
- All internet metrics continued to grow or maintained high levels from **2020 to 2022**.
- This suggests that pandemic-driven digital adoption has become permanent for many Canadians.
- **Voice calls** showed minimal growth from **2020 to 2022** (64% to 65.2%), indicating this behavior stabilized after the pandemic surge.

### Age Group Differences:

#### Older Canadians (65+):
- **Internet access**: +9.21% (the most significant increase among all groups).
- **Social media use**: +10% (the highest change across all groups).

#### Younger Demographics (15-24):
- **Internet access**: +1.02%.
- **Social media**: 0%.
- These groups showed minimal change, as they were already heavy users of both.

#### Middle-Aged Adults (45-64):
- Showed moderate but significant increases across all metrics.


## Midway Checkpoint

In [17]:
df = pd.read_csv("../data/merged_data_clean.csv")
df

age_group  year  access_to_internet  sent_received_emails  \
0             15 to 24 years  2018                98.6                  92.3   
1             25 to 44 years  2018                98.2                  94.2   
2             45 to 64 years  2018                93.9                  87.8   
3          65 years and over  2018                71.2                  62.9   
4   Total, 15 years and over  2018                91.3                  85.5   
5             15 to 24 years  2020                98.0                  90.5   
6             25 to 44 years  2020                98.2                  95.0   
7             45 to 64 years  2020                94.2                  87.3   
8          65 years and over  2020                76.3                  67.0   
9   Total, 15 years and over  2020                92.3                  85.9   
10            15 to 24 years  2022                99.2                  93.3   
11            25 to 44 years  2022                99.2                  96.2   
12            45 to 64 years  2022                96.4                  92.2   
13         65 years and over  2022                82.6                  74.7   
14  Total, 15 years and over  2022                94.5                  89.6   

    instant_messaging  social_media  voice_calls  dating_apps  news  research  \
0                89.6          91.1         68.1         11.0  71.4      88.4   
1                88.6          86.4         60.5          8.3  83.7      90.9   
2                68.6          62.8         40.1          2.6  73.1      82.6   
3                33.6          32.4         21.2          0.8  46.7      56.9   
4                71.1          68.5         47.0          5.3  71.0      80.9   
5                93.1          91.3         80.1         14.9  76.9       NaN   
6                90.9          87.2         79.6         10.8  89.4       NaN   
7                75.1          65.9         59.1          4.6  81.4       NaN   
8                44.1          39.7         37.2          1.8  57.1       NaN   
9                76.3          70.9         64.1          7.5  78.2       NaN   
10               92.1          90.5         81.8         14.8  75.9       NaN   
11               91.2          89.6         80.1         13.4  85.4       NaN   
12               78.2          72.3         61.0          6.7  80.0       NaN   
13               50.9          43.5         37.8          2.3  61.2       NaN   
14               78.0          73.7         64.7          8.9  76.9       NaN   

    ...  twenty_plus_stream_videos  twenty_plus_play_videos  work  school  \
0   ...                        NaN                      NaN   NaN     NaN   
1   ...                        NaN                      NaN   NaN     NaN   
2   ...                        NaN                      NaN   NaN     NaN   
3   ...                        NaN                      NaN   NaN     NaN   
4   ...                        NaN                      NaN   NaN     NaN   
5   ...                       25.4                     14.8  38.8    35.6   
6   ...                       20.3                      5.2  51.5     6.1   
7   ...                       11.9                      2.0  38.2     3.3   
8   ...                        8.5                      1.9   6.1     0.3   
9   ...                       15.8                      4.9  35.8     8.2   
10  ...                       25.9                     15.0  45.4    62.9   
11  ...                       21.4                      7.8  58.4    10.9   
12  ...                       13.7                      2.6  41.8     4.8   
13  ...                        8.8                      1.7   6.1     0.2   
14  ...                       16.8                      5.8  39.5    13.5   

    public_library  public_location  business  home  someone_else_home  \
0              NaN              NaN       NaN   NaN                NaN   
1              NaN              NaN       NaN   NaN         

In [47]:
df.shape

(15, 62)

In [48]:
df.info

<bound method DataFrame.info of                    age_group  year  access_to_internet  sent_received_emails  \
0             15 to 24 years  2018                98.6                  92.3   
1             25 to 44 years  2018                98.2                  94.2   
2             45 to 64 years  2018                93.9                  87.8   
3          65 years and over  2018                71.2                  62.9   
4   Total, 15 years and over  2018                91.3                  85.5   
5             15 to 24 years  2020                98.0                  90.5   
6             25 to 44 years  2020                98.2                  95.0   
7             45 to 64 years  2020                94.2                  87.3   
8          65 years and over  2020                76.3                  67.0   
9   Total, 15 years and over  2020                92.3                  85.9   
10            15 to 24 years  2022                99.2                  93.3   
11      

In [49]:
df.columns

Index(['age_group', 'year', 'access_to_internet', 'sent_received_emails',
       'instant_messaging', 'social_media', 'voice_calls', 'dating_apps',
       'news', 'research', 'health_research', 'events_research',
       'services_research', 'music', 'podcasts', 'video_streaming',
       'video_sharing_websites', 'subscription_video', 'sports', 'eSports',
       'video_game_stream', 'tv', 'books_magazines', 'play_video_games',
       'gambling', 'buy_online_boards', 'search_jobs', 'online_banking',
       'online_investing', 'virtual_wallet', 'online_formal_training',
       'informal_training', 'booked_appointments', 'track_health',
       'online_gov_services', 'smart_home', 'online_shopping', 'digital_goods',
       'physical_goods', 'peer_ride_services', 'peer_accomodation_services',
       'other', 'prop_use_internet', 'prop_stream_videos',
       'prop_play_video_games', 'less_than_ten_hours_internet',
       'less_than_ten_hrs_stream_video', 'less_than_ten_hrs_play_video_games',


In [50]:
internet_activities = ['access_to_internet', 'sent_received_emails', 'instant_messaging', 'social_media', 'voice_calls']
df_melted = pd.melt(df, 
                    id_vars=['age_group', 'year', 'New_cases'], 
                    value_vars=internet_activities,
                    var_name='activity_type', 
                    value_name='percentage')
df_melted


age_group  year     New_cases       activity_type  \
0             15 to 24 years  2018      0.000000  access_to_internet   
1             25 to 44 years  2018      0.000000  access_to_internet   
2             45 to 64 years  2018      0.000000  access_to_internet   
3          65 years and over  2018      0.000000  access_to_internet   
4   Total, 15 years and over  2018      0.000000  access_to_internet   
..                       ...   ...           ...                 ...   
70            15 to 24 years  2022  27326.269231         voice_calls   
71            25 to 44 years  2022  27326.269231         voice_calls   
72            45 to 64 years  2022  27326.269231         voice_calls   
73         65 years and over  2022  27326.269231         voice_calls   
74  Total, 15 years and over  2022  27326.269231         voice_calls   

    percentage  
0         98.6  
1         98.2  
2         93.9  
3         71.2  
4         91.3  
..         ...  
70        81.8  
71        80.1  
72        61.0  
73        37.8  
74        64.7  

[75 rows x 5 columns]

In [51]:
chart = alt.Chart(df_melted).mark_line(point=True).encode(
    x=alt.X('year:O', title='Year'),
    y=alt.Y('percentage:Q', title='Percentage of Population (%)'),
    color=alt.Color('age_group:N', title='Age Group'),
    column=alt.Column('activity_type:N', title='Internet Activity'),
    tooltip=['year:O', 'age_group:N', 'percentage:Q', 'New_cases:Q']
).properties(
    width=175,
    height=300,
    title='Changes in Internet Usage Across Age Groups (2018-2022)'
)
chart

alt.Chart(...)

##### Initial Observations 
 - Significant increase in usage of internet services in Seniors(>65 years).
 - Stagnated change in voice calls after 2020.

In [52]:
df_2018 = df[df['year'] == 2018].copy()
df_2020 = df[df['year'] == 2020].copy()
df_2022 = df[df['year'] == 2022].copy()

In [53]:
results = []

for age in df_2018['age_group'].unique():
    for activity in internet_activities:
        val_2018 = df_2018[df_2018['age_group'] == age][activity].values[0]
        val_2020 = df_2020[df_2020['age_group'] == age][activity].values[0]
        val_2022 = df_2022[df_2022['age_group'] == age][activity].values[0]
        
        # Percentage change
        change_2018_2020 = ((val_2020 - val_2018) / val_2018) * 100
        change_2020_2022 = ((val_2022 - val_2020) / val_2020) * 100
        
        results.append({
            'age_group': age,
            'activity': activity,
            'period': '2018-2020 (Early Pandemic)',
            'percent_change': change_2018_2020,
            'covid_cases': df_2020[df_2020['age_group'] == age]['New_cases'].values[0]
        })
        
        results.append({
            'age_group': age,
            'activity': activity,
            'period': '2020-2022 (Later Pandemic)',
            'percent_change': change_2020_2022,
            'covid_cases': df_2022[df_2022['age_group'] == age]['New_cases'].values[0]
        })

df_changes = pd.DataFrame(results)
df_changes

age_group              activity  \
0             15 to 24 years    access_to_internet   
1             15 to 24 years    access_to_internet   
2             15 to 24 years  sent_received_emails   
3             15 to 24 years  sent_received_emails   
4             15 to 24 years     instant_messaging   
5             15 to 24 years     instant_messaging   
6             15 to 24 years          social_media   
7             15 to 24 years          social_media   
8             15 to 24 years           voice_calls   
9             15 to 24 years           voice_calls   
10            25 to 44 years    access_to_internet   
11            25 to 44 years    access_to_internet   
12            25 to 44 years  sent_received_emails   
13            25 to 44 years  sent_received_emails   
14            25 to 44 years     instant_messaging   
15            25 to 44 years     instant_messaging   
16            25 to 44 years          social_media   
17            25 to 44 years          social_media   
18            25 to 44 years           voice_calls   
19            25 to 44 years           voice_calls   
20            45 to 64 years    access_to_internet   
21            45 to 64 years    access_to_internet   
22            45 to 64 years  sent_received_emails   
23            45 to 64 years  sent_received_emails   
24            45 to 64 years     instant_messaging   
25            45 to 64 years     instant_messaging   
26            45 to 64 years          social_media   
27            45 to 64 years          social_media   
28            45 to 64 years           voice_calls   
29            45 to 64 years           voice_calls   
30         65 years and over    access_to_internet   
31         65 years and over    access_to_internet   
32         65 years and over  sent_received_emails   
33         65 years and over  sent_received_emails   
34         65 years and over     instant_messaging   
35         65 years and over     instant_messaging   
36         65 years and over          social_media   
37         65 years and over          social_media   
38         65 years and over           voice_calls   
39         65 years and over           voice_calls   
40  Total, 15 years and over    access_to_internet   
41  Total, 15 years and over    access_to_internet   
42  Total, 15 years and over  sent_received_emails   
43  Total, 15 years and over  sent_received_emails   
44  Total, 15 years and over     instant_messaging   
45  Total, 15 years and over     instant_messaging   
46  Total, 15 years and over          social_media   
47  Total, 15 years and over          social_media   
48  Total, 15 years and over           voice_calls   
49  Total, 15 years and over           voice_calls   

                        period  percent_change   covid_cases  
0   2018-2020 (Early Pandemic)       -0.608519    759.057692  
1   2020-2022 (Later Pandemic)        1.224490  27326.269231  
2   2018-2020 (Early Pandemic)       -1.950163    759.057692  
3   2020-2022 (Later Pandemic)        3.093923  27326.269231  
4   2018-2020 (Early Pandemic)        3.906250    759.057692  
5   2020-2022 (Later Pandemic)       -1.074114  27326.269231  
6   2018-2020 (Early Pandemic)        0.219539    759.057692  
7   2020-2022 (Later Pandemic)       -0.876232  27326.269231  
8   2018-2020 (Early Pandemic)       17.621145    759.057692  
9   2020-2022 (Later Pandemic)        2.122347  27326.269231  
10  2018-2020 (Early Pandemic)        0.000000    759.057692  
11  2020-2022 (Later Pandemic)        1.018330  27326.269231  
12  2018-2020 (Early Pandemic)        0.849257    759.057692  
13  2020-2022 (Later Pandemic)        1.263158  27326.269231  
14  2018-2020 (Early Pandemic)        2.595937    759.057692  
15  2020-2022 (Later Pandemic)        0.330033  27326.269231  
16  2018-2020 (Early Pandemic)        0.925926    759.057692  
17  2020-2022 (Later Pandemic)        2.752294  27326.269231  
18  2018-2020 (Early Pandemic)       31.570248    759.057692  
19  

In [54]:
heatmap = alt.Chart(df_changes).mark_rect().encode(
    x=alt.X('activity:N', title='Internet Activity'),
    y=alt.Y('age_group:N', title='Age Group', sort=['15 to 24 years', '25 to 44 years', '45 to 64 years', '65 years and over']),
    color=alt.Color('percent_change:Q', scale=alt.Scale(scheme='blueorange', domain=[-10, 40]), title='% Change'),
    column=alt.Column('period:N', title='Time Period'),
    tooltip=['age_group:N', 'activity:N', 'percent_change:Q', 'covid_cases:Q']
).properties(
    width=150,
    height=100,
    title='Impact of Pandemic on Internet Usage Patterns by Age Group'
)
heatmap

alt.Chart(...)

In [55]:
text = alt.Chart(df_changes).mark_text(color='black', fontSize=9).encode(
    x=alt.X('activity:N'),
    y=alt.Y('age_group:N', sort=['15 to 24 years', '25 to 44 years', '45 to 64 years', '65 years and over']),
    text=alt.Text('percent_change:Q', format='.1f'),
    column=alt.Column('period:N')
)
text


alt.Chart(...)

##### Visualization Submission
The visualization created shows the impact of the pandemic on internet usage patterns across different age groups from 2018 to 2022. It displays:

 1.  A line chart showing trends in internet activities across years(2018-2022)
 2.  A heatmap showing percentage changes in internet activities during the early pandemic (2018-2020) and later pandemic periods (2020-2022)
 3.  Text annotations showing the exact percentage change for each activity and age group

##### Purpose and Suitability of the Visualization
This visualization is meant to examine:

1. How has Canadians' daily internet use changed with the prevalence of COVID-19 cases?
2. Which internet activities saw the most significant changes and whether certain age groups were more affected than others

The heatmap is particularly suitable for this analysis because:

- It allows for easy comparison across multiple variables (age groups, activities, time periods)
- Color intensity immediately highlights the most significant changes
- The percentage values provide precise quantitative information
- The column layout separates early pandemic effects from later pandemic trends

##### Follow-up Work Needed Based on EDA Results
Based on the EDA results, the following follow-up work should be done:

- Correlation analysis: Examine the correlation between COVID-19 case numbers and internet activity changes
- Deeper dive into specific activities: Further analyze the activities that showed the most dramatic changes (like voice calls and instant messaging)
- Regional analysis: Compare these trends across different geographic regions to see if patterns differ
- Investigate home internet usage: The data shows "home" internet usage was only tracked from 2020 onward. A follow-up study could examine how the shift to home-based work and education specifically impacted internet usage patterns.

### Q5
##### How EDA Results Affect Data Handling
The EDA results will affect how we handle the data in several key ways:

- Age-stratified analysis: The substantial differences between age groups suggest that any further analysis should be stratified by age rather than looking at overall population averages.
- Normalization of changes: When analyzing percentage changes, I need to account for the vastly different baseline levels between age groups (e.g., older adults started with much lower internet usage).
- COVID-19 case correlation: The analysis should incorporate COVID-19 case numbers as a potential explanatory variable for changes in internet usage.
- Data transformation: For any modeling efforts, the percentage data should be transformed to improve normality (such as logit transformation for percentage data).

This exploratory analysis reveals that the pandemic had a profound impact on internet usage patterns, with older adults experiencing the most dramatic increases, particularly in activities like voice calls and instant messaging. This suggests that the pandemic may have accelerated digital adoption among previously underrepresented groups.


## Final Milestone